### PR TITLE
Openocd support for stm32f3 Discovery, stm32f4 Discovery and nucelof4 boards

### PR DESCRIPTION
**Pull request by krady21**
Added stm32f3 , stm32f4 and nucleo boards to the board list with openocd support.

### CODE REVIEW

**Comment by bradjc** [comment]
Thank you for this PR! We have just been discussing that we need to add this to Tockloader.

To make this easier to scale, I would prefer to add just two boards `stm32discovery` and `nucleo`. Do you happen to know how sensitive the boards are to the workareasize? Can we choose the same value for all boards?

The other issue is the app flash address. We generally reserve 0x20000 bytes for the kernel. I think it would be easier to just standardize on that since I'm assuming the stm32 kernels will all fit in that size.

**Comment by krady21** [comment]
I agree that adding just stm32discovery and nucleo would definitely be the  more scalable solution. In regards to workareasize, we might be able to choose the same value for all nucleo boards, but I'm not sure about the discovery ones. I'll have to look into it.

**Comment by alexandruradovici** [comment]
> To make this easier to scale, I would prefer to add just two boards `stm32discovery` and `nucleo`. Do you happen to know how sensitive the boards are to the workareasize? Can we choose the same value for all boards?
> 
The two boards use different configuration files in openocd.cfg as they have different MCU (F3 vs F4). The stm32f412g is compatible with the nucleo board.

File: tockloader/board_interface.py [hunk 1]
```diff
@@ -74,6 +74,55 @@ class BoardInterface:
 		         'openocd_commands': {'program': 'jtagspi_program {{binary}} {address:#x};',
 		                              'read': 'jtagspi_read {{binary}} {address:#x} {length};',
 		                              'erase': 'flash fillb {address:#x} 0x00 512;'}},
+        'stm32f3discovery': {'description': 'STM32F303VCT6 Discovery MCU',
```
Comment (bradjc) [inline tockloader/board_interface.py:77]:
```suggestion
        'stm32f3discovery': {'description': 'STM32F3-based Discovery Boards',
```

File: tockloader/board_interface.py [hunk 2]
```diff
@@ -74,6 +74,55 @@ class BoardInterface:
...
+                             'apps_start_address': 0x08020000,
+                             'page_size': 2048,
+                             'openocd': 'external',
+                             'openocd_prefix': 'interface hla; \
+                                                hla_layout stlink; \
+                                                hla_device_desc "ST-LINK/V2-1"; \
+                                                hla_vid_pid 0x0483 0x374b; \
+                                                set WORKAREASIZE 0xC000; \
+                                                source [find target/stm32f3x.cfg];'},
+        'stm32f412gdiscovery': {'description': 'STM32F412G Discovery MCU',
```
Comment (bradjc) [inline tockloader/board_interface.py:88]:
```suggestion
        'stm32f4discovery': {'description': 'STM32F4-based Discovery Boards',
```

File: tockloader/board_interface.py [hunk 3]
```diff
@@ -74,6 +74,55 @@ class BoardInterface:
...
+		                 'apps_start_address': 0x08040000,
+		                 'page_size': 2048,
+		                 'openocd': 'external',
+		                 'openocd_prefix': 'interface hla; \
+                                            hla_layout stlink; \
+                                            hla_device_desc "ST-LINK/V2-1"; \
+                                            hla_vid_pid 0x0483 0x374b; \
+                                            set WORKAREASIZE 0x10000; \
+                                            source [find target/stm32f4x.cfg];'},
+		'nucleo': {'description': 'STM32 Nucleo development boards',
```
Comment (bradjc) [inline tockloader/board_interface.py:121]:
```suggestion
		'nucleof4': {'description': 'STM32f4-based Nucleo development boards',
```

File: tockloader/board_interface.py [hunk 4]
```diff
@@ -74,6 +74,55 @@ class BoardInterface:
...
+		                 'arch': 'cortex-m4',
+		                 'apps_start_address': 0x08040000,
+		                 'page_size': 2048,
+		                 'openocd': 'external',
+		                 'openocd_prefix': 'interface hla; \
+                                            hla_layout stlink; \
+                                            hla_device_desc "ST-LINK/V2-1"; \
+                                            hla_vid_pid 0x0483 0x374b; \
+                                            set WORKAREASIZE 0x10000; \
+                                            source [find target/stm32f4x.cfg];'},
```
Comment (bradjc) [inline tockloader/board_interface.py:120]:
```suggestion
```

**Comment by krady21** [comment]
I have applied your suggestions and cleaned up the commits.

**Comment by bradjc** [comment]
Looks great! Thanks for submitting this!

**Pull request merged**